alicloud/aliyunclient/aliyunclient.go

```diff
@@ -0,0 +1,726 @@
+	"github.com/aliyun/fc-go-sdk"
+	"github.com/denverdino/aliyungo/cdn"
+	"github.com/denverdino/aliyungo/common"
+	"github.com/denverdino/aliyungo/cs"
+	"github.com/denverdino/aliyungo/dns"
+	"github.com/denverdino/aliyungo/kms"
+	"github.com/denverdino/aliyungo/location"
+	"github.com/denverdino/aliyungo/ram"
+	"github.com/hashicorp/terraform/terraform"
+)
+
+// AliyunClient of aliyun
+type AliyunClient struct {
+	Region   common.Region
+	RegionId string
+	//In order to build ots table client, add accesskey and secretkey in aliyunclient temporarily.
+	AccessKey                    string
+	SecretKey                    string
+	SecurityToken                string
+	OtsInstanceName              string
+	accountIdMutex               sync.RWMutex
+	config                       *Config
+	accountId                    string
+	ecsconn                      *ecs.Client
+	essconn                      *ess.Client
+	rdsconn                      *rds.Client
+	vpcconn                      *vpc.Client
+	slbconn                      *slb.Client
+	ossconn                      *oss.Client
+	dnsconn                      *dns.Client
+	ramconn                      ram.RamClientInterface
+	csconn                       *cs.Client
+	cdnconn                      *cdn.CdnClient
+	kmsconn                      *kms.Client
+	otsconn                      *ots.Client
+	cmsconn                      *cms.Client
+	logconn                      *sls.Client
+	fcconn                       *fc.Client
+	cenconn                      *cbn.Client
+	pvtzconn                     *pvtz.Client
+	ddsconn                      *dds.Client
+	stsconn                      *sts.Client
+	rkvconn                      *r_kvstore.Client
+	tablestoreconnByInstanceName map[string]*tablestore.TableStoreClient
+	csprojectconnByKey           map[string]*cs.ProjectClient
+}
+
+type ApiVersion string
+
+const (
+	ApiVersion20140526 = ApiVersion("2014-05-26")
+	ApiVersion20160815 = ApiVersion("2016-08-15")
+	ApiVersion20140515 = ApiVersion("2014-05-15")
+)
+
+const businessInfoKey = "Terraform"
+
+var goSdkMutex = sync.RWMutex{} // The Go SDK is not thread-safe
+
+// Client for AliyunClient
+func (c *Config) Client() (*AliyunClient, error) {
+	err := c.loadAndValidate()
+	if err != nil {
+		return nil, err
+	}
+
+	return &AliyunClient{
+		config:                       c,
+		Region:                       c.Region,
+		RegionId:                     c.RegionId,
+		AccessKey:                    c.AccessKey,
+		SecretKey:                    c.SecretKey,
+		SecurityToken:                c.SecurityToken,
+		OtsInstanceName:              c.OtsInstanceName,
+		accountId:                    c.AccountId,
+		tablestoreconnByInstanceName: make(map[string]*tablestore.TableStoreClient),
+		csprojectconnByKey:           make(map[string]*cs.ProjectClient),
+	}, nil
+}
+
+func (client *AliyunClient) RunSafelyWithEcsClient(do func(*ecs.Client) (interface{}, error)) (interface{}, error) {
```

Comment:
This PR aims to solve GO SDK thread-unsafe problem. I think the prefix `RunSafelyWith` is redundant and it can be removed to simplify the client definition. If you want to stress thread-safe, adding some annotations is good choice.

Reply:
I propose to still keep "With" like "WithSomeClient", if not it would look like a weird getter.